pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+
+    return x.numpy()
+
+
+ar.register_function("autograd", "to_numpy", _to_numpy_autograd)
+
+
+def _scatter_element_add_autograd(tensor, index, value):
+    size = tensor.size
+    flat_index = _i("qml").numpy.ravel_multi_index(index, tensor.shape)
+    t = [0] * size
+    t[flat_index] = value
+    return tensor + _i("qml").numpy.array(t).reshape(tensor.shape)
+
+
+ar.register_function("autograd", "scatter_element_add", _scatter_element_add_autograd)
+
+
+def _take_autograd(tensor, indices, axis=None):
+    indices = _i("qml").numpy.asarray(indices)
+
+    if axis is None:
+        return tensor.flatten()[indices]
+
+    fancy_indices = [slice(None)] * axis + [indices]
+    return tensor[tuple(fancy_indices)]
+
+
+ar.register_function("autograd", "take", _take_autograd)
+
+
+# -------------------------------- TensorFlow --------------------------------- #
+
+
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "angle"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arcsin"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arccos"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arctan"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "diag"] = "tensorflow.linalg"
+
+
+ar.autoray._FUNC_ALIASES["tensorflow", "arcsin"] = "asin"
+ar.autoray._FUNC_ALIASES["tensorflow", "arccos"] = "acos"
+ar.autoray._FUNC_ALIASES["tensorflow", "arctan"] = "atan"
+ar.autoray._FUNC_ALIASES["tensorflow", "diag"] = "diag"
+
+
+ar.register_function("tensorflow", "asarray", lambda x: _i("tf").convert_to_tensor(x))
+ar.register_function("tensorflow", "flatten", lambda x: _i("tf").reshape(x, [-1]))
+ar.register_function("tensorflow", "shape", lambda x: tuple(x.shape))
+ar.register_function(
+    "tensorflow",
+    "sqrt",
+    lambda x: _i("tf").math.sqrt(
+        _i("tf").cast(x, "float64") if x.dtype.name in ("int64", "int32") else x
+    ),
+)
+
+
+def _take_tf(tensor, indices, axis=None):
+    tf = _i("tf")
+
+    if isinstance(indices, numbers.Number):
+        indices = [indices]
+
+    indices = tf.convert_to_tensor(indices)
+
+    if np.any(indices < 0):
+        # Unlike NumPy, TensorFlow doesn't support negative indices.
+        dim_length = tf.size(tensor).numpy() if axis is None else tf.shape(tensor)[axis]
+        indices = tf.where(indices >= 0, indices, indices + dim_length)
+
+    if axis is None:
+        # Unlike NumPy, if axis=None TensorFlow defaults to the first
+        # dimension rather than flattening the array.
+        data = tf.reshape(tensor, [-1])
+        return tf.gather(data, indices)
+
+    return tf.gather(tensor, indices, axis=axis)
+
+
+ar.register_function("tensorflow", "take", _take_tf)
+
+
+def _coerce_types_tf(tensors):
+    tf = _i("tf")
+    tensors = [tf.convert_to_tensor(t) for t in tensors]
+    dtypes = {i.dtype for i in tensors}
+
+    if len(dtypes) == 1:
+        return tensors
+
+    complex_type = dtypes.intersection({tf.complex64, tf.complex128})

```

Comment:
This part looks clean! Could perhaps the torch version be also written this way?